File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
    private static List<List<Integer>> tree;
    private static int[] parent;
    private static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        tree = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            tree.add(new ArrayList<>());
        }
        parent = new int[n + 1];
        visited = new boolean[n + 1];
        for (int i = 0; i < n - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            tree.get(a).add(b);
            tree.get(b).add(a);
        }

        dfs(1);

        StringBuilder sb = new StringBuilder();
        for (int i = 2; i <= n; i++) {
            sb.append(parent[i]).append('\n');
        }

        System.out.print(sb);
    }

    private static void dfs(int node) {
        visited[node] = true;

        for (int child : tree.get(node)) {
            if (!visited[child]) {
                parent[child] = node;
                dfs(child);
            }
        }
    }
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.kt
import java.io.BufferedReader
import java.io.InputStreamReader

private lateinit var tree: MutableList<MutableList<Int>>
private lateinit var parent: IntArray
private lateinit var visited: BooleanArray

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    tree = MutableList(n + 1) { mutableListOf() }
    parent = IntArray(n + 1)
    visited = BooleanArray(n + 1)
    repeat(n - 1) {
        val (a, b) = br.readLine().split(" ").map { it.toInt() }
        tree[a].add(b)
        tree[b].add(a)
    }

    dfs(1)

    val sb = StringBuilder()
    for (i in 2..n) {
        sb.append(parent[i]).append('\n')
    }

    print(sb)
}

private fun dfs(node: Int) {
    visited[node] = true

    for (child in tree[node]) {
        if (!visited[child]) {
            parent[child] = node
            dfs(child)
        }
    }
}
